### ISSUE TITLE
RankFeat: Rank-1 Feature Removal for Out-of-distribution Detection

### ISSUE THREAD

**Issue by zhuoxiao-chen** [Luoyadan/Research-Group-Progress#7]
### Paper Link
https://openreview.net/forum?id=-deKNiSOXLG
### OOD Detection

- A binary classiﬁcation problem with the goal to distinguish between ID and OOD data.
- The difﬁculty of OOD detection lies in designing an appropriate scoring function $S(·)$ such that the score distributions of ID and OOD data overlap as little as possible.

## Motivation

_The OOD feature tends to have a **much larger dominant singular value** than the ID feature, whereas the magnitudes of the **rest singular values are very similar**._

_Removing the rank-1 feature drastically perturbs the class prediction of OOD samples; a majority of predictions have been changed._

_On the contrary, most ID samples have consistent classiﬁcation results before and after removing the subspace._

_This phenomenon indicates that the **over-conﬁdent prediction of OOD samples** might be largely determined by the **dominant singular value** and the corresponding singular vectors._

<img width="1400" alt="image" src="https://user-images.githubusercontent.com/50613939/207499989-b2c0e5e6-5e59-49f1-a5a6-991a5d637ee7.png">

(a) The distribution of top-5 singular values for the ID and OOD features.
(b) Remove the rank-1 matrix composed by the dominant singular value and singular vectors, the class predictions of OOD data are severely perturbed, while those of ID data are moderately inﬂuenced.

<img width="1869" alt="image" src="https://user-images.githubusercontent.com/50613939/207501962-7082c4d8-a402-4a18-a124-88467fba07c3.png">

By removing the rank-1 feature, the over-conﬁdence of OOD samples is mitigated, and consequently the ID and OOD data can be better distinguished.

> FPR95: the false positive rate of OOD examples when the true positive rate for ID examples is 95%. The FPR95 threshold which is set when the true positive rate for ID examples is 95%.

### RankFeat: Rank-1 Feature Removal

1. SVD for high-level feature map: $\mathbf{X}=\mathbf{U S V}^T$
2. RankFeat removes the rank-1 matrix from the feature as $\mathbf{X}^{\prime}=\mathbf{X}-\mathbf{s}_1 \mathbf{u}_1 \mathbf{v}_1^T$, where $\mathbf{s}_1$ is the largest singular value
3. The perturbed feature is fed into the rest of the network to get logis prections $y'$
4. Energy score: $\operatorname{RankFeat}(\mathbf{x})=\log \sum \exp \left(\mathbf{y}^{\prime}\right)$

By removing the rank-1 matrix composed by the dominant singular value $s_1$ , the over-conﬁdent predictions of OOD data are largely perturbed. In contrast, the decisions of ID data are mildly inﬂuenced. This could help to separate the ID and OOD data better in the logit space.

### Acceleration by Power Iteration

Approximate the dominant eigenvector of a matrix

$$
\mathbf{v}_k=\frac{\mathbf{X} \mathbf{u}_k}{\left\|\mathbf{X} \mathbf{u}_k\right\|}, \mathbf{u}_{k+1}=\left(\frac{\mathbf{v}_k^T \mathbf{X}}{\left\|\mathbf{v}_k^T \mathbf{X}\right\|}\right)^T
$$

$\mathbf{u}_0$ and $\mathbf{v}_0$ are initialized with random orthogonal vectors and converge to the left and right singular vectors, respectively. The dominant singular value is computed as $\mathbf{s}_1=\mathbf{v}_k^T \mathbf{X} \mathbf{u}_k$. 

### Combination of Multi-scale Features (fusing the logits)

$$
\log \sum \exp \left(\left(\mathbf{y}^{\prime}+\mathbf{y}^{\prime \prime}\right) / 2\right)
$$

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/50613939/207578702-e512563c-ee67-4aa3-8ac6-42bb3151a346.png">

### Theory

**Proposition 1**. The upper bound of RankFeat score is defined as 

$$
\operatorname{RankFeat}(\mathbf{x})<\frac{1}{H W}\left(\sum_{i=1}^N \mathbf{s}_i-\mathbf{s}_1\right)\|\mathbf{W}\|_{\infty}+\|\mathbf{b}\|_{\infty}+\log (Q)
$$

where $Q$ denotes the number of categories, and $\mathbf{W}$ and $\mathbf{b}$ are the weight and bias of the last layer, respectively. A larger $\mathbf{s}_1$ would reduce the upper bound more.

**The upper bound** can be used to explain the shrinkage and skew of score distributions, $\operatorname{RankFeat}$ would squeeze score distributions.

<img width="1391" alt="image" src="https://user-images.githubusercontent.com/50613939/207574029-b98f6cb9-9fb1-4e47-9582-c1b532f81275.png">

Since the dominant singular value $\mathbf{s}_1$ contributes most to the score, removing $\mathbf{s}_1$ is likely to make many samples have similar scores. 

Given that the OOD feature tends to have a much larger $\mathbf{s}_1$, this would have a greater impact on OOD data and skew the OOD score distribution more.

#### Removing the rank-1 matrix is likely to make the statistics of OOD features closer to random matrices.

> For a random matrix of a given shape, the density of its eigenvalue asymptotically converges to the Manchenko-Pastur (MP) distribution. 

Compute the KL divergence between the actual eigenvalue distribution and the ﬁtted MP distribution.

<img width="800" alt="image" src="https://user-images.githubusercontent.com/50613939/207575848-347cf0da-87ac-48fc-8d27-5864988f1b17.png">

This indicates that removing the rank-1 matrix makes the statistics of OOD feature closer to random matrices, i.e., the OOD feature is very likely to become less informative than the ID feature. 

### SVD
https://stats.stackexchange.com/a/134283

### Power Iteration
> https://blog.csdn.net/qq_38402294/article/details/124413080

## Implementation 

- Mainly evaluate our method on the large-scale ImageNet1k benchmark.
- For the OOD datasets, we select four testsets from subsets of iNaturalist, SUN, Places, and Textures. These datasets are crafted by [26] with non-overlapping categories from ImageNet1k.
- Ablation on Google-BiT-S ResNetv2-101 model. 

## Experiments 

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/50613939/207577260-05784000-3f27-4c93-8267-78456070ff5a.png">

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/50613939/207577359-90d957e2-06f4-455c-9f76-921ca22d16ca.png">

### Ablation
<img width="1000" alt="image" src="https://user-images.githubusercontent.com/50613939/207578044-8b04b42d-e663-40ee-8cc1-b6604d5fa364.png">
